### PR TITLE
FOCUS #182: Usage Quantity

### PR DESCRIPTION
**Pull request by MichaelArkoosh-vega**
FOCUS-182-Usage-Quantity: Adding Usage Quantity Specification & Supporting Content

### CODE REVIEW

File: specification/dimensions/usagequantity.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
+# Usage Quantity
+
+Usage Quantity is the amount of resources or services consumed. The Usage Quantity is measured in units specified by the Usage Unit. Usage Quantity may differ from a Pricing Quantity when providers use a different unit of measure for calculating cost.Usage Quantity is often used when analyzing environments to gain understanding of usage/consumption of given products/services in an environment.
```
Comment (amitw-g) [inline specification/dimensions/usagequantity.md:3]:
```suggestion
Usage Quantity is the amount of resources or services consumed. The Usage Quantity is measured in units specified by the Usage Unit. Usage Quantity may differ from a Pricing Quantity when providers use a different unit of measure for calculating cost. Usage Quantity is often used when analyzing environments to gain understanding of usage or consumption of a given product/service in an environment.

```

**Review by amitw-g — APPROVED** [review]

File: specification/dimensions/usagequantity.md (hunk 1 above)
Comment (ljadvey) [inline specification/dimensions/usagequantity.md]: Usage Quantity is the amount of resources or services consumed. The Usage Quantity is measured in units specified by the Usage Unit. Usage Quantity may differ from a Pricing Quantity when providers use a different unit of measure for calculating cost. Usage Quantity is often used when analyzing environments to gain understanding of usage/consumption of given products/services in an environment.

**Review by ljadvey — COMMENTED** [review]
Looks good.

File: specification/dimensions/usagequantity.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
+# Usage Quantity
```
Comment (amitw-g) [inline specification/dimensions/usagequantity.md:1]: The name should be updated to QuantityInUsageUnit to be consistent with other names.

File: specification/dimensions/quantityinusageunit.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
+# Quantity In Usage Unit
+
+Quantity In Usage Unit is the amount of resources or services consumed. The Quantity In Usage Unit is measured in units specified by the Usage Unit. Quantity In Usage Unit may differ from a Quantity In Pricing Unit when providers use a different unit of measure for calculating cost. Quantity In Usage Unit is often used when analyzing environments to gain understanding of usage or consumption of a given product/service in an environment.
```
Comment (flanakin) [inline specification/dimensions/quantityinusageunit.md:3]: Can we generalize to be more inclusive?
```suggestion
Quantity In Usage Unit is the amount of a resource or service that was used or purchased. The Quantity In Usage Unit is measured in units specified by the Usage Unit. Quantity In Usage Unit may differ from a Quantity In Pricing Unit when providers use a different unit of measure for calculating cost. Quantity In Usage Unit is often used when analyzing environments to gain understanding of usage or consumption of a given product/service in an environment.
```

File: specification/dimensions/quantityinusageunit.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+Quantity In Usage Unit MUST exist in billing data. The column MUST be a numeric value and MUST NOT contain nulls.
+
+## Column ID
+
+quantity_in_usage_unit
```
Comment (flanakin) [inline specification/dimensions/quantityinusageunit.md:9]:
```suggestion
QuantityInUsageUnit
```

**Comment by tobrien** [comment]
@MichaelArkoosh-vega @ljadvey @flanakin @ijurica @marcin-walkow 

Related to the 181 PR branch - added in the paragraph that makes an explicit BCP-14 statement that this is usage and note pricing.

Also, removed the word "purchase" from this column definition, changed the value format specification, and added a description.

File: specification/dimensions/quantityinusageunit.md [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+Quantity In Usage Unit is the number of units of a resource or service that was used or consumed. The Quantity In Usage Unit is measured in units specified by the Usage Unit. Quantity In Usage Unit is used in conjunction with Usage Unit to gain understanding of usage or consumption of a given product/service in an environment.
+
+Quantity In Usage Unit MUST exist in billing data. The column MUST be a numeric value and MUST NOT contain nulls.
+
+The Quantity In Usage Unit for a particular line item may differ from Quantity In Pricing Unit when providers use different quantity, quantity increments, or columns to determine cost. The values in UsageUnit and QuantityInUsageUnit are often listed at a finer granularity or over a different time interval than the PricingUnit and QuantityInPricingUnit. While QuantityInUsageUnit may appear to relate to pricing and cost, QuantityInUsageUnit is focused on resource or service consumption. It is essential not to confuse QuantityInUsageUnit with QuantityInPricingUnit which is the basis for determining cost. The QuantityInUsageUnit column MUST not be used as the basis for determining values related to any pricing or cost metric.
```
Comment (mattray) [inline specification/dimensions/quantityinusageunit.md:7]: "MUST NOT"

File: specification/dimensions/quantityinusageunit.md [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+The Quantity In Usage Unit for a particular line item may differ from Quantity In Pricing Unit when providers use different quantity, quantity increments, or columns to determine cost. The values in UsageUnit and QuantityInUsageUnit are often listed at a finer granularity or over a different time interval than the [PricingUnit](#pricingunit) and [QuantityInPricingUnit](#quantityinpricingunit). While QuantityInUsageUnit may appear to relate to pricing and cost, QuantityInUsageUnit is focused on resource or service consumption. It is essential not to confuse QuantityInUsageUnit with QuantityInPricingUnit which is the basis for determining cost. The QuantityInUsageUnit column MUST NOT be used as the basis for determining values related to any pricing or cost metric.
+
+Quantity In Usage Unit MUST exist in billing data. The column MUST be a numeric value and when the [ChargeType](#chargetyep) is "Usage" Quantity In Usage Unit SHOULD NOT contain nulls. Negative values are only permissable when providers are adjusting for usage corrections.
```
Comment (TDubovchenko) [inline specification/dimensions/quantityinusageunit.md:7]: " Quantity In Usage Unit SHOULD NOT contain nulls" implies that it *may* contain null, however column is declared as not allowing nulls

File: specification/dimensions/quantityinusageunit.md [hunk 7]
```diff
@@ -0,0 +1,32 @@
+# Quantity In Usage Unit
```
Comment (flanakin) [inline specification/dimensions/quantityinusageunit.md:1]: As we've seen columns mature and close on a common prefixed grouping pattern, I would really like to see this and `QuantityInPricingUnit` be called `UsageQuantity` and `PricingQuantity` so they sit next to their respective unit columns in a list. These will be very out-of-place if not sitting next to the column that explains what the number means.

File: specification/dimensions/quantityinusageunit.md (hunk 6 above)
Comment (flanakin) [reply specification/dimensions/quantityinusageunit.md:7]: This column has to allow nulls for taxes and some adjustments. But I do agree this feels too soft. If we want this filled out, we should be explicit. Suggesting an update to be more complete and align with others.
```suggestion
QuantityInUsageUnit MUST exist in billing data. The column MUST be a numeric value. When [ChargeType](#chargetype) is "Usage" or "Purchase", QuantityInUsageUnit MUST NOT be null or empty. When ChargeType is "Adjustment" and the adjustment applies to specific charges that had a common UsageUnit, QuantityInUsageUnit MUST NOT be null. Negative values are only permissable when providers are adjusting for usage corrections.
```

File: specification/dimensions/quantityinusageunit.md [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+The number of units of a resource or service that was used or consumed based on the [UsageUnit](#usageunit)
+
+## Content constraints
+
+| Constraint      | Value        |
+|-----------------|--------------|
+| Column required | True         |
+| Data type       | Numeric      |
+| Allows nulls    | False        |
```
Comment (flanakin) [inline specification/dimensions/quantityinusageunit.md:27]:
```suggestion
| Allows nulls    | True         |
```

File: specification/dimensions/quantityinusageunit.md (hunk 5 above)
Comment (MichaelArkoosh-vega) [reply specification/dimensions/quantityinusageunit.md]: Corrected, thanks!

**Review by marcin-walkow — CHANGES_REQUESTED** [review]
Please use **PricingMeasure** (non-normalized) instead of PricingUnit (semi-normalized)

File: specification/dimensions/usagequantity.md [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+Usage Quantity represents the volume of a resource or service that was consumed, measured in units specified by the [UsageUnit](#usageunit). Distinct from [Pricing Quantity](#pricingquantity), Usage Quantity is used in conjunction with Usage Unit to measure the consumption of a product/service offering from a provider.
+
+The Usage Quantity for a charge may differ from Pricing Quantity when providers use different quantity, quantity increments, or columns to determine cost. The values in UsageUnit and UsageQuantity are often listed at a finer granularity or over a different time interval than the [PricingMeasure](#pricingmeasure) and [PricingQuantity](#pricingquantity). While UsageQuantity may appear to relate to pricing and cost, UsageQuantity is focused on resource consumption. It is essential not to confuse the UsageQuantity column with the PricingQuantity column which is the basis for determining cost. The UsageQuantity column MUST NOT be used as the basis for determining values related to any pricing or cost metric.
```
Comment (flanakin) [inline specification/dimensions/usagequantity.md:5]:
```suggestion
The Usage Quantity for a charge may differ from Pricing Quantity when providers use different quantity, quantity increments, or columns to determine cost. The values in UsageUnit and UsageQuantity are often listed at a finer granularity or over a different time interval than the [PricingMeasurementUnit](#pricingmeasurementunit) and [PricingQuantity](#pricingquantity). While UsageQuantity may appear to relate to pricing and cost, UsageQuantity is focused on resource consumption. It is essential not to confuse the UsageQuantity column with the PricingQuantity column which is the basis for determining cost. The UsageQuantity column MUST NOT be used as the basis for determining values related to any pricing or cost metric.
```

File: specification/dimensions/usagequantity.md [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+
+UsageQuantity
+
+## Display name
+
+Usage Quantity
+
+## Description
+
+The number of units of a resource or service that was used or consumed based on the [UsageUnit](#usageunit)
```
Comment (flanakin) [inline specification/dimensions/usagequantity.md:19]:
```suggestion
Number of units of a resource or service that was used or consumed based on the [UsageUnit](#usageunit).
```

**Review by flanakin — APPROVED** [review]

**Review by ljadvey — APPROVED** [review]

**Review by rileyjenk — APPROVED** [review]

File: specification/dimensions/usagequantity.md [hunk 11]
```diff
@@ -0,0 +1,31 @@
+# Usage Quantity
+
+The Usage Quantity represents the volume of a given resource and/or service used or purchased based on the [Usage Unit](#usageunit). Usage Quantity is often derived at a finer granularity or over a different time interval when compared to the [Pricing Quantity](#pricingquantity) (complementary to [Pricing Measurement Unit](#pricingmeasurementunit)), and focuses on resource and service consumption, not pricing and cost.
+
+UsageQuantity MUST be present in the billing data. This column MUST be a numeric value of type Decimal. The value MAY be negative in cases where [Charge Subcategory](#chargesubcategory) is 'Refund'. This column MUST NOT contain null values when [SKU Price ID](#skupriceid) is not null.
```
Comment (ijurica) [inline specification/dimensions/usagequantity.md:5]:
```suggestion
UsageQuantity MUST be present in the billing data. This column MUST be a numeric value of type Decimal. The value MAY be negative in cases where [ChargeSubcategory](#chargesubcategory) is 'Refund'. This column MUST NOT contain null values when [SkuPriceId](#skupriceid) is not null.
```

**Review by udam-f2 — APPROVED** [review]
LGTM

**Review by cnharris10 — APPROVED** [review]

**Review by ijurica — APPROVED** [review]

**Review by marcin-walkow — APPROVED** [review]

**Pull request merged**